### DISCUSSION TITLE
Interrupt pyexec_raw_repl() and pyexec_friendly_repl()

### DISCUSSION THREAD

**Discussion by JackAtKitronik** [micropython/micropython#9989] [Core Development]
Wondering if anyone has any ideas of how I might interrupt the pyexec_raw_repl() and pyexec_friendly_repl() from inside main.c. I would like the firmware to exit the pyexec_friendly_repl() when another control function is invoked but can't seem to figure out how to interrupt the REPL function. Here is the section of code from main.c in the rp2 port.

    for (;;) {
        if (pyexec_mode_kind == PYEXEC_MODE_RAW_REPL) {
            if (pyexec_raw_repl() != 0) {
                break;
            }
        } else {
            if (pyexec_friendly_repl() != 0) {
                break;
            }
        }
    }

**Comment by jimmo** [comment]
If I understand what you're asking, you'd like a way for something that was invoked from the REPL to cause the REPL to exit (similar to as if they'd hit Ctrl-D). I'm not sure what you mean by "from inside main.c" though.

I think the thing to look at is the handling of SystemExit in pyexec.c. By default SystemExit doesn't cause the REPL loop to terminate, but you can configure this with the `pyexec_system_exit` global variable. See the implementation of `machine.soft_reset()` which does (I think) exactly what you're asking. https://github.com/micropython/micropython/blob/master/ports/rp2/modmachine.c#L64

Note that user code can intercept SystemExit, but see https://github.com/micropython/micropython/pull/9949

**Reply by JackAtKitronik** [reply]
So I'm modifying the C code for the rp2 port to create my own RP2040 bootloader. I want to be able to connect to the MicroPython device using two different methods. One of these methods should be the standard REPL that can be used in programs like Thonny. Then I'm creating a second connection method. When I connect to the device using the second method, I want to interrupt and exit the standard REPL.

I have attempted using the implementation of `machine.soft_reset()` when I connect to the device using the second method but unfortunately this has no effect on the REPL in Thonny.

Thanks for your suggestions though.

**Reply by jimmo** [reply]
Sorry I still don't quite understand... Is your bootloader implemented in Python? Is it the bootloader that has a REPL?

> this has no effect on the REPL in Thonny.

If I understand you're saying you want the REPL to be disconnected? Like you want the serial port to close automatically?

**Reply by JackAtKitronik** [reply]
> Sorry I still don't quite understand... Is your bootloader implemented in Python? Is it the bootloader that has a REPL?

Okay, the MicroPython bootloader is implemented in C. Both the standard serial REPL and my REPL are also implemented in the C bootloader.

> If I understand you're saying you want the REPL to be disconnected? Like you want the serial port to close automatically?

Yes exactly. When a connection to the MicroPython device is made (which I can detect in the C bootloader) I want to close the standard serial port based REPL. This will in turn disconnect the MicroPython device from any IDE it is connected to, like Thonny. 

I would like to do this because the serial port based REPL is a blocking process and this stops my REPL from being able to run its processing loop in the C bootloader.

**Reply by jimmo** [reply]
Still not getting it sorry... what is "the MicroPython bootloader"?

Or are your saying that your bootloader has a REPL?

Or is your bootloader running concurrently with MicroPython? (Not really clear on how that's a bootloader though)

> This will in turn disconnect the MicroPython device from any IDE it is connected to, like Thonny.

I'm not aware of any device-initiated way to terminate a serial connection, except for a CDC ACM UART by tearing down the USB stack. (For example if you trigger a hard reset, this causes the /dev/ttyACM0 device to disasppear until the device re-enumerates).

**Reply by JackAtKitronik** [reply]
> I'm not aware of any device-initiated way to terminate a serial connection, except for a CDC ACM UART by tearing down the USB stack. (For example if you trigger a hard reset, this causes the /dev/ttyACM0 device to disasppear until the device re-enumerates).

Okay, no worries. Seems like I can't implement it in the way I was hoping to. Will have to figure out an alternative approach.

Thank you for your help.

**Reply by jimmo** [reply]
I'm not saying it's impossible, just that I don't know of any way to do it.

Have you seen any other devices do this (other than by triggering a reset or USB re-enumeration?)

**Reply by JackAtKitronik** [reply]
> Have you seen any other devices do this (other than by triggering a reset or USB re-enumeration?)

No I haven't. And unfortunately, while a hard reset does work at closing the serial connection, this isn't a valid option as it also disconnects my REPL connection.
